File: core/PublicCode.php
<?php

require_once __DIR__ . '/Database.php';
require_once dirname(__DIR__) . '/exceptions/AppException.php';

class PublicCode
{
    public const LENGTH = 12;

    public static function require(?string $value, string $field = 'code'): string
    {
        $value = strtolower(trim((string)$value));
        if ($value === '') {
            throw new AppException(400, 'MISSING_FIELD', "Missing URL parameter: {$field}");
        }

        if (!preg_match('/^[a-f0-9]{12}$/', $value)) {
            throw new AppException(400, 'INVALID_CODE', 'Invalid code format');
        }

        return $value;
    }

    public static function generate(): string
    {
        return bin2hex(random_bytes(intdiv(self::LENGTH, 2)));
    }

    public static function generateUnique(string $table): string
    {
        if (!in_array($table, ['tb_kungfus', 'tb_tasks'], true)) {
            throw new AppException(500, 'INTERNAL_ERROR', 'Invalid code target');
        }

        $db = Database::getInstance();
        for ($attempt = 0; $attempt < 10; $attempt++) {
            $code = self::generate();
            $exists = $db->fetchOne(
                "SELECT 1 FROM {$table} WHERE code = :code LIMIT 1",
                [':code' => $code]
            );
            if (!$exists) {
                return $code;
            }
        }

        throw new AppException(500, 'CODE_GENERATION_FAILED', 'Could not generate a unique code');
    }
}

File: internal/publiccode/publiccode.go
package publiccode

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"regexp"
	"strings"
)

const Length = 12

var codePattern = regexp.MustCompile(`^[a-f0-9]{12}$`)

// Require validates a code parameter. Returns error if invalid.
func Require(value *string, field string) (string, error) {
	if value == nil {
		return "", fmt.Errorf("MISSING_FIELD: Missing URL parameter: %s", field)
	}
	v := *value
	normalized := toLower(trim(v))
	if normalized == "" {
		return "", fmt.Errorf("MISSING_FIELD: Missing URL parameter: %s", field)
	}
	if !codePattern.MatchString(normalized) {
		return "", fmt.Errorf("INVALID_CODE: Invalid code format")
	}
	return normalized, nil
}

// Generate creates a random 12-hex-char code.
func Generate() string {
	b := make([]byte, 6)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

// GenerateUnique generates a code that doesn't exist in the given table.
// The checkFunc should query the database to see if the code exists.
func GenerateUnique(checkFunc func(string) (bool, error)) (string, error) {
	for attempt := 0; attempt < 10; attempt++ {
		code := Generate()
		exists, err := checkFunc(code)
		if err != nil {
			return "", err
		}
		if !exists {
			return code, nil
		}
	}
	return "", fmt.Errorf("CODE_GENERATION_FAILED: Could not generate a unique code")
}

func toLower(s string) string { return strings.ToLower(s) }
func trim(s string) string    { return strings.TrimSpace(s) }
